sail/impl/plugin.py

```diff
@@ -0,0 +1,394 @@
+
+
+P = typing.ParamSpec("P")
+T = typing.TypeVar("T")
+
+_CommandMapping = typing.Dict[str, command_trait.Command[typing.Any, typing.Any]]
+_ListenerMapping = typing.Dict[
+    typing.Type[base_events.Event],
+    typing.List[base_events.EventCallbackT[base_events.Event]],
+]
+_EntrypointCallback = typing.Callable[
+    [velum.GatewayClient, command_manager_trait.CommandManager],
+    typing.Coroutine[typing.Any, typing.Any, None],
+]
+
+
+_LOGGER = logging.getLogger("sail.plugin")
+_UNIONS = frozenset((typing.Union, types.UnionType))
+
+
+def _get_source_module_name() -> str:
+    module_name = pathlib.Path(logging.currentframe().f_code.co_filename).stem
+    _LOGGER.debug("Module name resolved to %r", module_name)
+    return module_name
+
+
+class Plugin(plugin_trait.Plugin):
+
+    name: str
+    logger: logging.Logger
+
+    _client: typing.Optional[velum.GatewayClient]
+    _command_manager: typing.Optional[command_manager_trait.CommandManager]
+    _commands: _CommandMapping
+    _listeners: _ListenerMapping
+
+    def __init__(
+        self, name: typing.Optional[str] = None, logger: typing.Optional[logging.Logger] = None
+    ) -> None:
+        self.name = name or _get_source_module_name()
+        self.logger = _LOGGER if logger is None else logger
+
+        self._client = None
+        self._command_manager = None
+        self._commands = {}
+        self._listeners = {}
+
+    @property
+    def client(self) -> velum.GatewayClient:
+        if self._client is None:
+            raise RuntimeError("Cannot access the client of an inactive plugin.")
+
+        return self._client
+
+    @property
+    def rest(self) -> rest_trait.RESTClient:
+        return self.client.rest
+
+    @property
+    def event_manager(self) -> event_manager_trait.EventManager:
+        return self.client.event_manager
+
+    @property
+    def command_manager(self) -> command_manager_trait.CommandManager:
+        if self._command_manager is None:
+            raise RuntimeError("Cannot access the command manager of an inactive plugin.")
+
+        return self._command_manager
+
+    @property
+    def commands(self) -> typing.Sequence[command_trait.Command[typing.Any, typing.Any]]:
+        return tuple(self._commands.values())
+
+    @property
+    def listeners(
+        self,
+    ) -> typing.Mapping[
+        typing.Type[base_events.Event],
+        typing.Sequence[base_events.EventCallbackT[base_events.Event]],
+    ]:
+        return {event: tuple(listeners) for event, listeners in self._listeners.items()}
+
+    def add_command(self, command: command_trait.Command[typing.Any, typing.Any]) -> None:
+        if command.name in self._commands:
+            raise RuntimeError(f"Plugin {self.name} already has a command named {command.name!r}.")
+
+        self._commands[command.name] = command
+
+    def remove_command(self, command: command_trait.Command[typing.Any, typing.Any]) -> None:
+        if command.name in self._commands:
```

Comment:
```suggestion
        if command.name not in self._commands:
```
Also this method doesn't actually remove the command.

Reply:
Thanks, fixed in 8239f7201c4956e615727d734ece0ad18d4ec075.